File: lib/attribution.ml
(** See [Attribution.mli] for documentation. *)

type origin = Det | NonDet

let string_of_origin = function Det -> "det" | NonDet -> "nondet"

type outcome =
  | Passed
  | Policy_failed of { reason : string }
  | Transition_blocked of {
      from_state : string;
      to_state : string;
      reason : string;
    }
  | Partial_pass of { score : float; rationale : string }

type t = {
  origin : origin;
  gate : string;
  evidence : Yojson.Safe.t;
  outcome : outcome;
}

(* --- Serialization --- *)

let outcome_to_yojson : outcome -> Yojson.Safe.t = function
  | Passed -> `Assoc [ ("kind", `String "passed") ]
  | Policy_failed { reason } ->
    `Assoc [ ("kind", `String "policy_failed"); ("reason", `String reason) ]
  | Transition_blocked { from_state; to_state; reason } ->
    `Assoc
      [
        ("kind", `String "transition_blocked");
        ("from_state", `String from_state);
        ("to_state", `String to_state);
        ("reason", `String reason);
      ]
  | Partial_pass { score; rationale } ->
    `Assoc
      [
        ("kind", `String "partial_pass");
        ("score", `Float score);
        ("rationale", `String rationale);
      ]

let to_yojson (t : t) : Yojson.Safe.t =
  `Assoc
    [
      ("origin", `String (string_of_origin t.origin));
      ("gate", `String t.gate);
      ("evidence", t.evidence);
      ("outcome", outcome_to_yojson t.outcome);
    ]

(* --- Smart constructors --- *)

let passed ~origin ~gate ~evidence = { origin; gate; evidence; outcome = Passed }

let policy_failed ~origin ~gate ~evidence ~reason =
  { origin; gate; evidence; outcome = Policy_failed { reason } }

let transition_blocked ~origin ~gate ~evidence ~from_state ~to_state ~reason =
  {
    origin;
    gate;
    evidence;
    outcome = Transition_blocked { from_state; to_state; reason };
  }

let partial_pass ~origin ~gate ~evidence ~score ~rationale =
  { origin; gate; evidence; outcome = Partial_pass { score; rationale } }

File: dashboard/src/api/attribution.ts
// MASC Dashboard — Attribution REST client (Layer 4).
//
// Reads from the in-process ring buffer in `lib/dashboard/dashboard_attribution.ml`
// via /api/v1/attribution/{recent,summary}. Plain types — no valibot schema yet
// because the envelope is frozen at the OCaml side (Attribution.to_yojson) and
// the dashboard is the only consumer.

import { get, type GetOptions } from './core'

export type AttributionOrigin = 'det' | 'nondet'

export type AttributionOutcome =
  | { kind: 'passed' }
  | { kind: 'policy_failed'; reason: string }
  | { kind: 'transition_blocked'; from_state: string; to_state: string; reason: string }
  | { kind: 'partial_pass'; score: number; rationale: string }

export interface Attribution {
  origin: AttributionOrigin
  gate: string
  evidence: unknown
  outcome: AttributionOutcome
}

export interface AttributionEvent {
  attribution: Attribution
  recorded_at: number
}

interface AttributionRecentResponse {
  events: AttributionEvent[]
  count: number
}

export interface GateSummary {
  gate: string
  passed: number
  policy_failed: number
  transition_blocked: number
  partial_pass: number
  total: number
}

export interface AttributionSummaryResponse {
  gates: GateSummary[]
}

export async function fetchAttributionSummary(
  opts: GetOptions = {},
): Promise<AttributionSummaryResponse> {
  return get<AttributionSummaryResponse>('/api/v1/attribution/summary', opts)
}

export async function fetchAttributionRecent(
  params: { gate?: string; limit?: number } = {},
  opts: GetOptions = {},
): Promise<AttributionRecentResponse> {
  const sp = new URLSearchParams()
  if (params.gate) sp.set('gate', params.gate)
  if (params.limit !== undefined) sp.set('limit', String(params.limit))
  const query = sp.toString()
  const path = query
    ? `/api/v1/attribution/recent?${query}`
    : '/api/v1/attribution/recent'
  return get<AttributionRecentResponse>(path, opts)
}
